vnext/Microsoft.ReactNative/Views/TextViewManager.cpp

```diff
@@ -55,7 +68,53 @@ class TextShadowNode final : public ShadowNodeBase {
       Super::AddView(*m_firstChildNode, 0);
       m_firstChildNode = nullptr;
     }
+
     Super::AddView(child, index);
+
+    if (auto run = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Run>()) {
+      if (m_isHighlighted) {
+        AddHighlighter(m_ColorValue, run.Text().size());
+      }
+
+      m_prevCursorEnd += run.Text().size();
+    } else if (auto span = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Span>()) {
+      AddNestedTextHighlighter(
+          m_isHighlighted, m_ColorValue, span, static_cast<VirtualTextShadowNode &>(child).m_highlightData.get());
+    }
+  }
+
+  void AddNestedTextHighlighter(
+      bool isParentHighlighted,
+      folly::dynamic parentColor,
+      winrt::Span &span,
+      VirtualTextShadowNode::HighlightData *highData) {
+    if (!highData->isHighlighted && isParentHighlighted) {
+      highData->isHighlighted = true;
+      highData->color = parentColor;
+    }
+
+    for (const auto &el : span.Inlines()) {
+      if (auto run = el.try_as<winrt::Run>()) {
+        if (highData->isHighlighted) {
+          AddHighlighter(highData->color, run.Text().size());
+        }
+
+        m_prevCursorEnd += run.Text().size();
+      } else if (auto spanChild = el.try_as<winrt::Span>()) {
+        AddNestedTextHighlighter(
+            highData->isHighlighted, highData->color, spanChild, highData->data[highData->spanIdx++].get());
+      }
+    }
+  }
+
+  void AddHighlighter(const folly::dynamic &color, size_t runSize) {
+    auto newHigh = winrt::TextHighlighter{};
+    newHigh.Background(SolidColorBrushFrom(color));
+
```

Comment:
Normally with highlighters you also want to ensure that the text being highlighted is visible so that highlighting a section changes the text foreground color such that a minimum contrast ratio is achieved.

Reply:
Interesting! Does XAML do it automatically in any scenario? If not, where can I find guidance on what color I should change foreground to?
Also, can't we rely on the developer to take care of this? If they specify a weird highlighter color, they can also supply a different text color alongside it, perhaps conditionally in JS.

Reply:
It looks like the TextHighlighter feature doesn't do it automatically since it's a lower level primitive, so I guess my question is whether the RN API should do it or leave it up to the developer. It's expected our controls comply with WCAG (see [here](https://webaim.org/articles/contrast/#:~:text=WCAG%20requires%20%22at%20least%204.5%3A1%22%20contrast%2C%20so%20you,It%20does%20not%20meet%20the%20WCAG%20contrast%20threshold.) but it isn't clear to me whether that means that we automatically set foreground, or if we just allow and trust that the developer do the right thing (TM)

Reply:
Seems like we should match existing RN behavior. Maybe we could test with an Expo Snack quickly?

Reply:
Tried the following style in a Snack {backgroundColor: #777777", color: 'gray'} - #777777 is metnioned in the WCAG doc as non-compliant. Looks like RN doesn't adjust anything.
![image](https://user-images.githubusercontent.com/12816515/96942830-1299a180-148b-11eb-8992-446e8fa2b332.png)